### ISSUE TITLE
the repeat report problem of using 'self.log' in pytorch_lightning.LightningModule

### ISSUE THREAD

**Issue by qiwang1996** [Lightning-AI/lightning#18128]
### Bug description

hello, i run my training code on 20 devices and  there are 4 GPUs  for each one, that is, i start 80 training processes  totally. Moreover, i use  IterableDataset. I set  my log-function like this: 
self.log("train_loss", loss.item(),  on_epoch=False, prog_bar=True, logger=True, rank_zero_only=True)

 and i got  the  log output like this:

Epoch 0: : 2it [00:47, 23.73s/it, loss=0.178, v_num=11, train_loss=0.178]
Epoch 0: : 2it [00:47, 23.73s/it, loss=0.178, v_num=11, train_loss=0.178]
Epoch 0: : 2it [00:47, 23.73s/it, loss=0.161, v_num=11, train_loss=0.145]
Epoch 0: : 3it [00:54, 18.23s/it, loss=0.161, v_num=11, train_loss=0.145]
Epoch 0: : 3it [00:54, 18.23s/it, loss=0.161, v_num=11, train_loss=0.145]
Epoch 0: : 3it [00:54, 18.23s/it, loss=0.153, v_num=11, train_loss=0.135]

 However, i want  each line to log gathered training loss information of each step over all training process.   It may seem like this:

Epoch 0: : 2it [00:47, 23.73s/it, loss=0.178, v_num=11, train_loss=0.178]
Epoch 0: : 3it [00:54, 18.23s/it, loss=0.161, v_num=11, train_loss=0.145]

How can i do to  solve this and why  are there three lines for  each step  as log information currently?  

### What version are you seeing the problem on?

v1.6

### Error messages and logs

```
# Error messages and logs here please
```
 

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

cc @carmocca @Blaizzy

**Comment by awaelchli** [comment]
The repeated log-lines you see might be because you are using Lightning 1.6 which is very old and this was probably fixed a while ago. If you still see this problem on 1.9 or later, could you please provide a code example? Thanks!

**Comment by qiwang1996** [comment]
thanks , i will follow your advice and try the latest lightning version.  If there is still  a repeated problem,  i will report it.

**Comment by awaelchli** [comment]
I assume this was solved. But let me know if there is still an issue.

**Comment by xor-xor** [comment]
> FYI, I see that you set self.log(rank_zero_only=True). Make sure not to set this, it is for a different purpose.

@awaelchli could you expand a bit what do you mean by "for a different purpose"? To my understaning, when I have something like this in my distributed training:

```python
def validation_epoch_end(self, outputs):
    metrics = self.val_metrics.compute()  # self.val_metrics is a MetricCollection from TorchMetrics
    self.log_dict(metrics, rank_zero_only=True)
    self.val_metrics.reset()
```

...then it is totally fine to use `rank_zero_only=True`, because at this point `metrics` is just a dict, and all synchronization already happened in `.compute()`, so there's no risk of a deadlock.

(and yes, I know that I can log `self.val_metrics` directly as a metrics object, but let's narrow this to a situation where I have to call `self.val_metrics.compute()` because of some reasons which are irrelevant here)

(and for a further background why I need `rank_zero_only=True`, see [this comment](https://github.com/Lightning-AI/lightning/issues/17037#issuecomment-1670024952))

**Comment by awaelchli** [comment]
@xor-xor The docs for this argument says:

> rank_zero_only: Whether the value will be logged only on rank 0. This will prevent synchronization which
>                 would produce a deadlock as not all processes would perform this log call.

A code example will make it clear. `rank_zero_only` needs to be set to `True` by the user if their intention is to make the log call *explicitly* only on rank 0. Like this:

```py
if self.global_rank == 0:
    self.log_dict(metrics, rank_zero_only=True)  # <--- setting this is mandatory, otherwise program hangs!
```

This is how the argument is designed to be used but 99.9% of users never need to ever set this or even look at this.
It does *not* control whether the result gets written to the logger on rank 0 or on all ranks and this is what my initial comment

> FYI, I see that you set self.log(rank_zero_only=True). Make sure not to set this, it is for a different purpose.

was saying, because it looked like the user thought multiple lines were printed despite setting this argument. But again, I was pointing out that these are unrelated to each other.

**Comment by xor-xor** [comment]
@awaelchli thank you for a good explanation! 👍🏻 I'd consider adding it to the documentation (along with your remark that probably 99.9% of users never need to touch it) because it really clarifies the purpose of this argument.
